vmagent fails to build if the configuration contains environment variables

## Describe the bug

I'm using environment variables in vmagent configuration to access secrets (via [sops-nix](https://github.com/Mic92/sops-nix)).
Since ad06fc9 the build fails since the dry-run environment doesn't contain the environment variables set at runtime.

## Steps To Reproduce

Steps to reproduce the behavior:

1. use this testcase
```
{
  ...
}:
{
  imports = [ ./hardware-configuration.nix ];
  boot.loader.systemd-boot.enable = true;

  services.vmagent = {
    enable = true;
    prometheusConfig = {
      scrape_configs = [
        {
          job_name = "vmagent";
          static_configs = [ { targets = [ "localhost:8429" ]; } ];
          authorization = {
            type = "Bearer";
            credentials = "%{TOKEN}";
          };
        }
      ];
    };
  };
  systemd.services.vmagent.serviceConfig.EnvironmentFile = ./example.env;

  system.stateVersion = "25.05";
}
```
2. Build with `nixos-rebuild build -I nixos-config=./configuration.nix`

## Expected behavior

The build is successful

## Screenshots

```
error: builder for '/nix/store/9qmn0c7gf4z8p682v0bqdrylzbjinnpj-checked-config.drv' failed with exit code 255;
       last 5 log lines:
       > 2024-12-22T20:15:04.874Z       info    VictoriaMetrics/lib/logger/flag.go:12   build version: 1.108.0
       > 2024-12-22T20:15:04.874Z        info    VictoriaMetrics/lib/logger/flag.go:13   command-line flags
       > 2024-12-22T20:15:04.874Z    info    VictoriaMetrics/lib/logger/flag.go:20     -dryRun="true"
       > 2024-12-22T20:15:04.874Z      info    VictoriaMetrics/lib/logger/flag.go:20     -promscrape.config="/nix/store/qanwxx2gwbyb9rdfz8fyqwm64v7nmwk1-prometheusConfig.yaml"
       > 2024-12-22T20:15:04.874Z      fatal   VictoriaMetrics/app/vmagent/main.go:118 error when checking -promscrape.config: cannot parse Prometheus config from "/nix/store/qanwxx2gwbyb9rdfz8fyqwm64v7nmwk1-prometheusConfig.yaml": cannot unmarshal data: cannot expand environment variables: missing "TOKEN" env var
       For full logs, run 'nix log /nix/store/9qmn0c7gf4z8p682v0bqdrylzbjinnpj-checked-config.drv'.
```

## Additional context

It's nice to have configuration validation at build time, but it's tricky to do when using runtime environment variables.

I see maybe two options:
1. flag to disable validation (which would restore pre ad06fc9 behavior)
2. ability to inject "fake" environment variables when performing the dry-run.

## Metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.61, NixOS, 25.05 (Warbler), 25.05.20241115.5e4fbfb`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.10`
 - nixpkgs: `/nix/store/hfz1qqd0z8amlgn8qwich1dvkmldik36-source`

## Notify maintainers

@ryan4yin

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/nixos/modules/services/databases/victoriametrics.nix
+++ b/nixos/modules/services/databases/victoriametrics.nix
@@ -24,10 +24,13 @@ let
 
   checkedConfig =
     file:
-    pkgs.runCommand "checked-config" { nativeBuildInputs = [ cfg.package ]; } ''
-      ln -s ${file} $out
-      ${lib.escapeShellArgs startCLIList} -promscrape.config=${file} -dryRun
-    '';
+    if cfg.checkConfig then
+      pkgs.runCommand "checked-config" { nativeBuildInputs = [ cfg.package ]; } ''
+        ln -s ${file} $out
+        ${lib.escapeShellArgs startCLIList} -promscrape.config=${file} -dryRun
+      ''
+    else
+      file;
 in
 {
   options.services.victoriametrics = {
@@ -125,6 +128,17 @@ in
         Extra options to pass to VictoriaMetrics. See the docs:
         <https://docs.victoriametrics.com/single-server-victoriametrics/#list-of-command-line-flags>
         or {command}`victoriametrics -help` for more information.
+      '';
+    };
+
+    checkConfig = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = ''
+        Check configuration.
+
+        If you use credentials stored in external files (`environmentFile`, etc),
+        they will not be visible  and it will report errors, despite a correct configuration.
       '';
     };
   };

--- a/nixos/modules/services/monitoring/vmagent.nix
+++ b/nixos/modules/services/monitoring/vmagent.nix
@@ -30,10 +30,13 @@ let
 
   checkedConfig =
     file:
-    pkgs.runCommand "checked-config" { nativeBuildInputs = [ cfg.package ]; } ''
-      ln -s ${file} $out
-      ${lib.escapeShellArgs startCLIList} -promscrape.config=${file} -dryRun
-    '';
+    if cfg.checkConfig then
+      pkgs.runCommand "checked-config" { nativeBuildInputs = [ cfg.package ]; } ''
+        ln -s ${file} $out
+        ${lib.escapeShellArgs startCLIList} -promscrape.config=${file} -dryRun
+      ''
+    else
+      file;
 in
 {
   imports = [
@@ -117,6 +120,17 @@ in
         Extra args to pass to `vmagent`. See the docs:
         <https://docs.victoriametrics.com/vmagent.html#advanced-usage>
         or {command}`vmagent -help` for more information.
+      '';
+    };
+
+    checkConfig = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = ''
+        Check configuration.
+
+        If you use credentials stored in external files (`environmentFile`, etc),
+        they will not be visible  and it will report errors, despite a correct configuration.
       '';
     };
   };

